### PR TITLE
Polishing/project setup branch changes

### PR DESCRIPTION
**Pull request by aryann010**

### CODE REVIEW

File: Assets/Scripts/UI/UIView.cs [hunk 1]
```diff
@@ -29,10 +30,10 @@ private void Start()
         public void IncrementScore(int scoreToIncrement)
         {
             currentScore += scoreToIncrement;
-            scoreText.SetText(currentScore.ToString());
+            scoreText.text = currentScore.ToString();
```
Comment (mridul-outscal) [inline Assets/Scripts/UI/UIView.cs:44]: technically there is no difference here in both the approaches. But we will always prefer the one which is more readable in general. Hence we will be using SetText() method although internally its the same thing.

File: Assets/Scripts/PowerUps/PowerUps/Shield.cs [hunk 2]
```diff
@@ -11,12 +11,14 @@ public override void Activate()
         {
             base.Activate();
             GameService.Instance.GetPlayerService().GetPlayerController().SetShieldState(ShieldState.Activated);
+            GameService.Instance.GetPlayerService().GetPlayerController().ToggleShieldUI(true);
         }
```
Comment (mridul-outscal) [inline Assets/Scripts/PowerUps/PowerUps/Shield.cs:15]: The method ToggleShieldUI indicates it will turn shield on if its off already and vice-versa.
SetShieldUI/SetShieldState is a more apropriate name.
using enums over booleans everytime. even if its a flag like shield on/ shield off, we will create an enum for it and use that instead of a boolean.
Enums are much more cleaner and readable.
The code of these projects you are making is going to be read thousand of people, so always have readability on back of your mind while writing code.

File: Assets/Scripts/PowerUps/PowerUps/RapidFire.cs [hunk 3]
```diff
@@ -8,12 +8,14 @@ public override void Activate()
         {
             base.Activate();
             GameService.Instance.GetPlayerService().GetPlayerController().ToggleRapidFire(true);
+            GameService.Instance.GetMainMenu().ToggleRapidFireText(true);
         }
```
Comment (mridul-outscal) [inline Assets/Scripts/PowerUps/PowerUps/RapidFire.cs:12]: Read comment on Shield.cs

File: Assets/Scripts/PowerUps/PowerUps/DoubleTurret.cs [hunk 4]
```diff
@@ -8,12 +8,14 @@ public override void Activate()
         {
             base.Activate();
             GameService.Instance.GetPlayerService().GetPlayerController().ToggleDoubleTurret(true);
+            GameService.Instance.GetMainMenu().ToggleDoubleTurretText(true);
         }
```
Comment (mridul-outscal) [inline Assets/Scripts/PowerUps/PowerUps/DoubleTurret.cs:12]: Read comment on Shield.cs

File: Assets/Scripts/Player/PlayerView.cs [hunk 5]
```diff
@@ -14,7 +15,10 @@ public class PlayerView : MonoBehaviour, IDamageable
 
         public void SetController(PlayerController playerController) => this.playerController = playerController;
 
-        private void Update() => playerController.HandlePlayerInput();
+        private void Update()
+        {
+            playerController.HandlePlayerInput();
```
Comment (mridul-outscal) [inline Assets/Scripts/Player/PlayerView.cs:20]: We will always prefer lambda methods if it is single line of code.

File: Assets/Scripts/Player/PlayerScriptableObject.cs [hunk 6]
```diff
@@ -7,6 +7,7 @@ public class PlayerScriptableObject : ScriptableObject
     {
         public int maxHealth;
         public float movementSpeed;
+        public float rotationSpeed;
```
Comment (mridul-outscal) [inline Assets/Scripts/Player/PlayerScriptableObject.cs:10]: Why is this needed if the player is constantly looking towards the mouse position?

File: Assets/Scripts/Player/PlayerController.cs [hunk 7]
```diff
@@ -66,10 +66,13 @@ private void HandlePlayerMovement()
...
         {
-            // Rotate the player to look in the direction of mouse position.
+            // Rotate the player to look in the direction of the mouse position.
             var dir = Input.mousePosition - Camera.main.WorldToScreenPoint(playerView.transform.position);
             var angle = Mathf.Atan2(dir.y, dir.x) * Mathf.Rad2Deg;
-            playerView.transform.rotation = Quaternion.AngleAxis(angle - 90, Vector3.forward);
+
+            // Calculate the desired rotation based on the rotation speed and time.deltaTime
+            var desiredRotation = Quaternion.AngleAxis(angle - 90, Vector3.forward);
+            playerView.transform.rotation = Quaternion.RotateTowards(playerView.transform.rotation, desiredRotation, playerScriptableObject.rotationSpeed * Time.deltaTime);
```
Comment (mridul-outscal) [inline Assets/Scripts/Player/PlayerController.cs:75]: Seems flawed. from Quaternion.AngleAxis(angle - 90, Vector3.forward) we get the exact rotation for the player ship we need for any frame. SO we would want to set that rotation on every frame.
The player rotation speed depends on simply how fast you move the mouse.

File: Assets/Scripts/New/MainMenu.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
```
Comment (mridul-outscal) [inline Assets/Scripts/New/MainMenu.cs:6]: If the UI logic need these many serialized references, we will need to take a different approach than a simple UI Manager. We will have to implement MVC for each UI Screen. I have done this for the state machine project and the command pattern project. Sync with me to understand & implement this logic.

File: Assets/Scripts/GameService.cs [hunk 9]
```diff
@@ -8,12 +8,15 @@
...
 using CosmicCuration.Utilities;
-using CosmicCuration.PowerUps; 
+using CosmicCuration.PowerUps;
+using MainMenuUI;
 #endregion
 
 
 public class GameService : GenericMonoSingleton<GameService>
 {
+    public DifficultyLevel difficultyLevel;
```
Comment (mridul-outscal) [inline Assets/Scripts/GameService.cs:18]: What is DifficultyLevel?

File: Assets/Scripts/GameService.cs [hunk 10]
```diff
@@ -72,7 +76,35 @@ private void Update()
 
     public SoundService GetSoundService() => soundService;
 
-    public UIView GetUIService() => uiService; 
+    public UIView GetUIService() => uiService;
+
+    public MainMenu GetMainMenu() => mainMenu;
     #endregion
 
+    public void SetDifficultyVariables()
```
Comment (mridul-outscal) [inline Assets/Scripts/GameService.cs:89]: This is not the correct location in code for this method. Ideally if many components are getting affected by the difficulty level in the game, we should be having a separate DifficultyService which contains all the difficulty level logic.

File: Assets/Scripts/UI/UIView.cs [hunk 11]
```diff
@@ -43,6 +50,15 @@ public void EnableGameOverUI()
...
 
+        public void StoreHighScore()
+        {
+            if (currentScore > highScore)
+            {
+                highScore = currentScore;
+                PlayerPrefs.SetInt("HighScore", highScore);
+            }
+        }
+
```
Comment (mridul-outscal) [inline Assets/Scripts/UI/UIView.cs:72]: UI Service should not be the one responsible for saving data in Player Prefs, We will be abiding by Single Responsibility very strictly. We will create another Service for this functionality of needed, but not going to happen in UI.

File: Assets/Scripts/GameService.cs [hunk 12]
```diff
@@ -53,6 +53,11 @@ private void Start()
     {
         // Initialize all Services.
         soundService = new SoundService(soundScriptableObject, audioEffectSource, backgroundMusicSource);
+    }
+
+    public void Initialize()
+    {
+        SetDifficultyVariables();
         playerService = new PlayerService(playerPrefab, playerScriptableObject, playerBulletPrefab, playerBulletScriptableObject);
```
Comment (mridul-outscal) [inline Assets/Scripts/GameService.cs:61]: All Services will be initialized in Start() no need for Initialize() or SetDifficultyVariables() as it is not GameService's responsibility.

File: Assets/Scripts/New/MainMenu.cs [hunk 13]
```diff
@@ -82,6 +82,7 @@ private void Awake()
         private void OnClickStartButton()
         {
             homeScreen.SetActive(false);
+            GameService.Instance.Initialize();
             gameplayPanel.SetActive(true);
```
Comment (mridul-outscal) [inline Assets/Scripts/New/MainMenu.cs:86]: Nope, A method called Initialize() in GameService is being called from some other Script, this is not a good thing, Game Service is the ignition for all other Services.

File: Assets/Scripts/Audio/SoundService.cs [hunk 14]
```diff
@@ -9,16 +9,24 @@ public class SoundService
         private AudioSource audioEffects;
         private AudioSource backgroundMusic;
 
+        public bool isSoundEnabled;
+        public bool isMusicEnabled;
+
```
Comment (mridul-outscal) [inline Assets/Scripts/Audio/SoundService.cs:14]: Use Enums instead for SoundState if really needed. Don't use booleans.

File: Assets/Scripts/Audio/SoundService.cs [hunk 15]
```diff
@@ -9,16 +9,24 @@ public class SoundService
...
+            isSoundEnabled = true;
+            isMusicEnabled = true;
         }
 
         public void PlaySoundEffects(SoundType soundType, bool loopSound = false)
         {
+            if (!isSoundEnabled)
+                return;
+
             AudioClip clip = GetSoundClip(soundType);
```
Comment (mridul-outscal) [inline Assets/Scripts/Audio/SoundService.cs:30]: Enum check will be here rather than the bool check

File: Assets/Scripts/New/MainMenu.cs [hunk 16]
```diff
@@ -124,12 +124,12 @@ private void OnClickAudioPanelBackButton()
 
         private void OnClickSoundButton()
         {
-            
+            GameService.Instance.GetSoundService().SetSoundState();
         }
```
Comment (mridul-outscal) [inline Assets/Scripts/New/MainMenu.cs:128]: One line methods should always be lambda methods.

File: Assets/Scripts/Player/PlayerScriptableObject.cs (hunk 6 above)
Comment (aryann010) [reply Assets/Scripts/Player/PlayerScriptableObject.cs:10]: correct
